storage.bs

```diff
@@ -94,37 +100,292 @@ function retrieveNextChunk(nextChunkInfo) {
+<a>environment settings object</a> <var>environment</var>, and a <a>storage type</a>
+<var>type</var>, run these steps:
+
+<ol>
+ <li><p>Let <var>key</var> be <var>environment</var>'s
+ <a for="environment settings object">origin</a>.
+
+ <li><p>If <var>key</var> is an <a>opaque origin</a>, then return failure.
+
+ <li><p>If the user has disabled storage, then return failure.
+
+ <li>
+  <p>If <var>shed</var>[<var>key</var>] does not <a for=map>exist</a>, then:
+
+  <ol>
+   <li><p>Let <var>shelf</var> be a new <a>storage shelf</a>.
+
+   <li><p>Set <var>shelf</var>'s <a>bucket map</a>["<code>default</code>"] to the result of
+   <a>create a storage bucket</a> with <var>type</var>.
+
+   <li><p>Set <var>shed</var>[<var>key</var>] to <var>shelf</var>.
+  </ol>
+
+ <li><p>Return <var>shed</var>[<var>key</var>].
+</ol>
+
+<p>To <dfn>obtain a local storage shelf</dfn>, given an <a>environment settings object</a>
+<var>environment</var>, return the result of running <a>obtain a storage shelf</a> with the user
+agent's <a for="user agent">storage shed</a>, <var>environment</var>, and "<code>local</code>".
+
+
+<h3 id=buckets oldids=boxes>Storage buckets</h3>
 
-<p><a>Storage</a> consists of zero or more <dfn export id=site-storage-unit>storage units</dfn>.
+<p>A <dfn>storage bucket</dfn> is a place for <a>storage endpoints</a> to store data. Whenever a
+<a>storage bucket</a> is cleared by the user agent, it must be cleared in its entirety.
+<!-- This is probably better defined in some kind of storage lifetime section. -->
 
-<p>Each <a for=/>origin</a> has an associated <a>storage unit</a>. A <a>storage unit</a> contains a
-single <dfn export id=bucket oldids=box>bucket</dfn>. [[HTML]]
+<p>A <a>storage bucket</a> has a <dfn>bottle map</dfn> of <a>storage identifiers</a> to
+<a>storage bottles</a>.
 
+<hr>
 
-<h3 id=buckets oldids=boxes>Buckets</h3>
+<p>A <dfn id=bucket oldids=box>local storage bucket</dfn> is a <a>storage bucket</a> for local
+storage APIs.
 
-<p>A <a>bucket</a> has <dfn export for=bucket oldids=box-mode>mode</dfn> which is either
-"<code title>best-effort</code>" or "<code title>persistent</code>". A
-<dfn export oldids=persistent-box>persistent bucket</dfn> is a <a>bucket</a> whose
-<a for=bucket>mode</a> is "<code title>persistent</code>". A
-<dfn export oldids=non-persistent-box>non-persistent bucket</dfn> is a <a>bucket</a> whose
-<a for=bucket>mode</a> is <em>not</em> "<code title>persistent</code>".
+<p>A <a>local storage bucket</a> has a
+<dfn for="local storage bucket" id=bucket-mode oldids=box-mode>mode</dfn>, which is
+"<code>best-effort</code>" or "<code>persistent</code>". It is initially "<code>best-effort</code>".
 
-<p>A <a>bucket</a> is considered to be an atomic unit. Whenever a <a>bucket</a> is cleared by the
-user agent, it must be cleared in its entirety.
+<hr>
+
+<p>A <dfn>session storage bucket</dfn> is a <a>storage bucket</a> for session storage APIs.
```

Comment:
Can we have definitions and cross-references for "local storage" and "session storage"? Maybe

> A user agent holds a storage shed, which is a storage shed. All **local storage** data will be stored in this shed.